feat: add conversations_unreads and conversations_mark tools

saoudrizwan opened the pull request:
Adds two new tools for efficiently managing unread Slack messages 🚀

When you're in a lot of Slack channels (especially with external partner channels), keeping up with unreads becomes overwhelming. The existing `conversations_history` tool requires you to know which channel to check. These new tools let you:

1. Get a prioritized view of ALL unread channels in one API call
2. Filter to only channels where you're @mentioned
3. Mark channels as read after reviewing them

## New Tools

### `conversations_unreads`

Efficiently retrieves unread messages across all channels using a single API call.

**Parameters:**
| Parameter | Type | Default | Description |
|-----------|------|---------|-------------|
| `include_messages` | bool | `true` | Return actual messages or just channel summary |
| `channel_types` | string | `"all"` | Filter: `"dm"`, `"group_dm"`, `"partner"`, `"internal"`, or `"all"` |
| `max_channels` | int | `50` | Limit number of channels returned |
| `max_messages_per_channel` | int | `10` | Messages to fetch per channel |
| `mentions_only` | bool | `false` | Only return channels where you have @mentions |

**Output (CSV):**
```csv
ChannelID,ChannelName,ChannelType,UnreadCount,LastRead,Latest
D09EDMKJC64,@john,dm,5,,1760563532.527859
C08KX4PV3S6,#ext-partner-company,partner,2,1763493994.468259,1766596058.035729
C09J740AZTR,#engineering,internal,1,1759517692.989219,1762753562.005619
```

### `conversations_mark`

Marks a channel or DM as read.

**Parameters:**
| Parameter | Type | Required | Description |
|-----------|------|----------|-------------|
| `channel_id` | string | Yes | Channel ID, `#channel-name`, or `@username` |
| `ts` | string | No | Timestamp to mark read up to. If omitted, marks all as read |

## Technical Implementation

### Why `client.counts` API?

We initially tried using `ClientUserBoot` but discovered it only returns channels in the user's sidebar, not all channels with unreads. The `client.counts` API (already implemented in the edge client) is what Slack's web client uses:

- Returns `HasUnreads` boolean and `MentionCount` for ALL channels in one call
- Separates channels into three arrays: `Channels`, `MPIMs` (group DMs), `IMs` (direct messages)
- Much more efficient than checking each channel individually

### Channel Categorization & Priority

Results are automatically sorted by priority:
1. **DMs** (`dm`) - Direct messages, highest priority
2. **Group DMs** (`group_dm`) - Multi-person direct messages  
3. **Partner channels** (`partner`) - Externally shared channels (uses `IsExtShared` metadata)
4. **Internal channels** (`internal`) - Everything else

### Code Changes

**`pkg/provider/api.go`**
- Added `IsExtShared` field to `Channel` struct
- Exposed `ClientCounts` method on `SlackAPI` interface
- Updated `mapChannel` to pass through `IsExtShared`

**`pkg/handler/conversations.go`**
- Added `UnreadChannel` struct for CSV output
- Added `ConversationsUnreadsHandler` with priority sorting and filtering
- Added `ConversationsMarkHandler` with channel name resolution
- Added `sortChannelsByPriority` helper

**`pkg/server/server.go`**
- Registered both new tools with proper annotations

## Example Use Cases

### 1. Morning Inbox Review
```
"What are my unread messages?"
→ conversations_unreads (returns prioritized list: DMs first, then partner channels, then internal)
```

### 2. Priority Inbox - Just @Mentions  
```
"Show me only channels where I'm @mentioned"
→ conversations_unreads with mentions_only: true
```

### 3. Check Partner Channels Only
```
"What's new in external partner channels?"
→ conversations_unreads with channel_types: "partner"
```

### 4. Quick Summary Without Messages
```
"Which channels have unreads?"
→ conversations_unreads with include_messages: false
```

### 5. Mark Channel as Read
```
"Mark #random as read"
→ conversations_mark with channel_id: "#random"
```

### 6. Triage Workflow
```
1. Get unreads summary: conversations_unreads with include_messages: false
2. Review a specific channel: conversations_history
3. Mark it as read: conversations_mark
4. Repeat
```

## Breaking Changes

Users with existing channel caches will need to delete `~/Library/Caches/slack-mcp-server/channels_cache_v2.json` (or equivalent) to repopulate with the new `IsExtShared` field. Otherwise, partner channel detection won't work until the cache is refreshed.

## Test Plan

- [x] `conversations_unreads` returns prioritized list of unread channels
- [x] `channel_types` filter correctly filters by dm/group_dm/partner/internal
- [x] `mentions_only: true` filters to only channels with @mentions
- [x] `include_messages: false` returns summary without fetching message content
- [x] Partner channels detected via `IsExtShared` metadata (not name prefix)
- [x] `conversations_mark` marks channels as read using channel ID
- [x] `conversations_mark` resolves `#channel-name` to channel ID
- [x] `conversations_mark` resolves `@username` to DM channel ID
- [x] `conversations_mark` without timestamp marks all messages as read

korotovsky commented on pkg/handler/conversations.go:
```diff
@@ -341,6 +342,335 @@ func (ch *ConversationsHandler) ConversationsSearchHandler(ctx context.Context,
+			continue
+		}
+
+		// Get channel info from cache to determine type and name
```
I think would be better to reuse existing helper, it can handle more specific cases like groups and so on: https://github.com/korotovsky/slack-mcp-server/blob/master/pkg/handler/conversations.go#L711-L728

korotovsky replied:
OK let's keep this as is for now, if someone would report issues around it we will refactor it then.

korotovsky reviewed (commented):
Hi @saoudrizwan, thank you for the great contribution!

Since the new API method has been introduced and this MCP supports multiple slack token types, a question from my side: have you checked that `ClientCounts` also returns data if we provide `xoxb` or `xoxp` (as I assume your main setup is `xoxc`/`xoxd` pair since you are using undocumented APIs, this is is where this API works the best and has the most coverage)

saoudrizwan commented on pkg/handler/conversations.go:
```diff
@@ -341,6 +342,335 @@ func (ch *ConversationsHandler) ConversationsSearchHandler(ctx context.Context,
+	ch.logger.Debug("ConversationsUnreadsHandler called", zap.Any("params", request.Params))
+
+	// Get optional parameters
+	includeMessages := request.GetBool("include_messages", true)
```
Done in 4ffa603 - extracted to `unreadsParams` struct with `parseParamsToolUnreads` function.

saoudrizwan commented on pkg/handler/conversations.go:
```diff
@@ -341,6 +342,335 @@ func (ch *ConversationsHandler) ConversationsSearchHandler(ctx context.Context,
+}
+
+// ConversationsMarkHandler marks a channel as read up to a specific timestamp
+func (ch *ConversationsHandler) ConversationsMarkHandler(ctx context.Context, request mcp.CallToolRequest) (*mcp.CallToolResult, error) {
```
Done in 4ffa603 - added `SLACK_MCP_MARK_TOOL` environment variable. The tool is now disabled by default and requires explicit opt-in (`SLACK_MCP_MARK_TOOL=true` or `SLACK_MCP_MARK_TOOL=1`). Also updated the README with docs.

saoudrizwan commented on pkg/handler/conversations.go:
```diff
@@ -341,6 +342,335 @@ func (ch *ConversationsHandler) ConversationsSearchHandler(ctx context.Context,
+}
+
+// getChannelDisplayName returns a human-readable channel name
+func (ch *ConversationsHandler) getChannelDisplayName(id, name string, isIM, isMpIM bool, members []string, usersMap map[string]slack.User, channelsMaps *provider.ChannelsCache) string {
```
Removed in 4ffa603.

saoudrizwan commented:
Hey @korotovsky, thanks for the review! I've addressed all the inline comments in 4ffa603:

- Extracted params to structs with parsing functions (`unreadsParams`, `markParams`)
- Added `SLACK_MCP_MARK_TOOL` env var guard (disabled by default)
- Removed the unused `categorizeChannel` and `getChannelDisplayName` functions
- Updated README with the new env var docs

Regarding token compatibility: you're right that I've been testing with `xoxc`/`xoxd` tokens. The `client.counts` API is an undocumented endpoint that the Slack web client uses, so it's designed for session tokens. I haven't tested with `xoxb` or `xoxp` tokens.

Looking at the edge client implementation, `ClientCounts` just passes whatever token is configured. For `xoxb`/`xoxp` tokens, I'd expect one of these outcomes:
1. Works normally (unlikely since it's a web client API)
2. Returns an error like "not_allowed_token_type"
3. Returns empty/limited data

If you'd like, I can add a note in the README that `conversations_unreads` works best with `xoxc`/`xoxd` tokens, or we could add a token type check that falls back to a different approach for bot/OAuth tokens. Let me know what you'd prefer!
